### PR TITLE
Refresh ID Card inside Worker Thread to not be impacted by the load on the main Event Loop 

### PR DESCRIPTION
**Pull request by Shiranuit**
⚠️ Depends on #2145 

## What does this PR do ?

This PR exports the ID Card renewal mechanism of the IDCardHandler outside to be used inside a Worker Thread instead of the same Event Loop where Kuzzle is running.

### Why ?
Since a setInterval was used prior to this PR to refresh the ID Card periodically inside Redis, when the Kuzzle event loop is saturated because of the load, the refresh of the ID Cards takes too much time and the node will kill itself because it could not refresh his ID Card in time.

Now this mechanism is done in his own Worker Thread that has his own Event Loop.

### How should this be manually tested?

`npm run  test`

### CODE REVIEW

File: lib/cluster/workers/IDCardRenewer.js [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    await this.initRedis(config.redis);
+
+    /**
+     * Early renew, otherwise we need to wait the refreshDelay before the first ID Card renewal
+     * which could be enough to make the ID Card expire.
+     */
+    try {
+      await this.renewIDCard(); 
+    } catch (error) {
+      // Do nothing
```
Comment (alexandrebouthinon) [inline lib/cluster/workers/IDCardRenewer.js:34]: At least log something about the Redis error that could happend in the `renewCard` function

**Review by alexandrebouthinon — CHANGES_REQUESTED** [review]

**Review by xbill82 — APPROVED** [review]
👏

**Review by alexandrebouthinon — APPROVED** [review]

File: lib/cluster/idCardHandler.js [hunk 2]
```diff
@@ -99,28 +100,42 @@ class ClusterIdCardHandler {
...
-          // Unable to refresh the key in time before it expires
-          // => this node is too slow, we need to remove it from the cluster
-          if (refreshed === 0) {
-            await this.node.evictSelf('Node too slow: ID card expired');
-          }
-        }
+    this.refreshWorker = await this._constructWorker(`${__dirname}/workers/IDCardRenewer.js`);
+    this.refreshWorker.unref();
+
+    this.refreshWorker.on('message', async (message) => {
```
Comment (Aschen) [inline lib/cluster/idCardHandler.js:106]: Nit
```suggestion
    this.refreshWorker.on('message', async message => {
```

File: lib/cluster/idCardHandler.js [hunk 3]
```diff
@@ -99,28 +100,42 @@ class ClusterIdCardHandler {
...
+        config: global.kuzzle.config.services.internalCache,
+        name: 'internal_adapter',
       },
-      this.refreshDelay);
+      refreshDelay: this.refreshDelay,
+    });
+  }
+
+  // Used to Mock the creation of a worker for the tests
+  async _constructWorker(path) {
```
Comment (Aschen) [inline lib/cluster/idCardHandler.js:130]: Why this is async?

File: lib/cluster/workers/IDCardRenewer.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    this.refreshDelay = config.refreshDelay;
+    
+    /**
+    * Since we do not have access to the Kuzzle Context we can't use kuzzle.ask('core:cache:internal:*',...),
+    * so we need to have an instance of Redis similar to the one used in the Cache Engine
+    */
+    try {
+      await this.initRedis(config.redis);
+    } catch (error) {
+      this.parentPort.postMessage({
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.js:28]: nit
```suggestion
    } 
    catch (error) {
      this.parentPort.postMessage({
```

File: lib/cluster/workers/IDCardRenewer.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+          this.nodeIdKey,
+          this.refreshDelay * 1.5);
+        // Unable to refresh the key in time before it expires
+        // => this node is too slow, we need to remove it from the cluster
+        if (refreshed === 0) {
+          await this.dispose();
+          this.parentPort.postMessage({ error: 'Node too slow: ID card expired' });
+        }
+      }
+    } catch (error) {
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.js:62]: Wrong catch indentation

File: lib/cluster/workers/IDCardRenewer.js [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+  }
+
+  async dispose() {
+    if (!this.disposed) {
+      this.disposed = true;
+      clearInterval(this.refreshTimer);
+      this.refreshTimer = null;
+      try {
+        await this.redis.commands.del(this.nodeIdKey);
+      } catch (error) {
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.js:77]: Wrong catch indentation

File: lib/cluster/workers/IDCardRenewer.js [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+      global.kuzzle.log = {
+        debug: console.debug, // eslint-disable-line no-console
+        error: console.error, // eslint-disable-line no-console
+        info: console.info, // eslint-disable-line no-console
+        warn: console.warn, // eslint-disable-line no-console
+      };
+      
+      idCardRenewer.init(message);
+    } else if (message.action === 'dispose') {
+      idCardRenewer.dispose();
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.js:100]: I know it's not very useful but you should still attach a `catch` handler to the returned promise

**Review by Aschen — CHANGES_REQUESTED** [review]

File: lib/cluster/idCardHandler.js (hunk 3 above)
Comment (Shiranuit) [reply lib/cluster/idCardHandler.js:130]: No reason 😅

File: lib/cluster/workers/IDCardRenewer.js [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+      });
+      return;
+    }
+
+    /**
+     * Early renew, otherwise we need to wait the refreshDelay before the first ID Card renewal
+     * which could be enough to make the ID Card expire.
+     */
+    await this.renewIDCard();
+    if (!this.disposeed) {
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.js:40]: Typo
```suggestion
    if (!this.disposed) {
```

**Review by Aschen — CHANGES_REQUESTED** [review]

File: lib/cluster/workers/IDCardRenewer.js (hunk 8 above)
Comment (Shiranuit) [reply lib/cluster/workers/IDCardRenewer.js:40]: good catch 😅

File: lib/cluster/workers/IDCardRenewer.ts [hunk 9]
```diff
@@ -1,19 +1,24 @@
-'use strict';
-const { isMainThread, parentPort } = require('worker_threads');
-const Redis = require('../../service/cache/redis');
+import { isMainThread, parentPort } from 'worker_threads';
+import Redis from '../../service/cache/redis';
+import { JSONObject } from 'kuzzle-sdk';
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.ts:3]: Nit: import order
```suggestion
import { isMainThread, parentPort } from 'worker_threads';

import { JSONObject } from 'kuzzle-sdk';

import Redis from '../../service/cache/redis';
```

**Review by Aschen — APPROVED** [review]

**Comment by Shiranuit** [comment]
Waiting for [Wait index cache propagation before allowing node to process requests](#2145) to be ready before merging this PR

File: lib/cluster/workers/IDCardRenewer.ts (hunk 9 above)
Comment (Aschen) [reply lib/cluster/workers/IDCardRenewer.ts:3]: (It's still not the good order)

File: lib/cluster/workers/IDCardRenewer.ts [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+
+export class IDCardRenewer {
+  private disposed = false;
+  private redis : Redis = null;
+  private refreshTimer : ReturnType<typeof setInterval>;
+  private nodeIdKey : string = null;
+  private refreshDelay = 2000;
+  private parentPort : any = null;
```
Comment (Aschen) [inline lib/cluster/workers/IDCardRenewer.ts:11]: Why do you need to keep an instance reference on the global `parentPort` variable?

**Review by Aschen — CHANGES_REQUESTED** [review]

File: lib/cluster/workers/IDCardRenewer.ts (hunk 10 above)
Comment (Shiranuit) [reply lib/cluster/workers/IDCardRenewer.ts:11]: Just to be able to mock it in the tests

**Review by Aschen — APPROVED** [review]

**Pull request merged**